src/library/rssscanner.c

```diff
@@ -0,0 +1,612 @@
+
+  return 0;
+}
+
+static void
+mfi_metadata_fixup(struct media_file_info *mfi, struct rss_item_info *ri, const char *feed_title, const char *feed_author, uint32_t time_added)
+{
+  struct tm tm;
+
+  // Always take the meta from media file if possible; some podcasts (Apple) can
+  // use mp4 streams which tend not to have decent tags so  in those cases take
+  // info from the RSS and not the stream
+  if (!mfi->artist)
+    mfi->artist = safe_strdup(feed_author);
+  if (!mfi->album)
+    mfi->album  = safe_strdup(feed_title);
+  if (!mfi->url)
+    mfi->url    = safe_strdup(ri->link);
+  if (!mfi->genre || strcmp("(186)Podcast", mfi->genre) == 0)
+    {
+      free(mfi->genre);
+      mfi->genre  = strdup("Podcast");
+    }
+
+  // Title not valid on most mp4 (it becomes the url obj) so take from RSS feed
+  if (ri->type && strncmp("video", ri->type, 5) == 0)
+    {
+      free(mfi->title);
+      mfi->title = safe_strdup(ri->title);
+    }
+
+  // Remove, some can be very verbose
+  free(mfi->comment);
+  mfi->comment = NULL;
+
+  // Date is always from the RSS feed info
+  rss_date(&tm, ri->pubdate);
+  mfi->date_released = mktime(&tm);
+  mfi->year = 1900 + tm.tm_year;
+
+  mfi->media_kind = MEDIA_KIND_PODCAST;
+
+  mfi->time_added = time_added;
+}
+
+static int
+rss_save(struct playlist_info *pli, int *count, enum rss_scan_type scan_type)
+{
+  mxml_node_t *xml;
+  const char *feed_title;
+  const char *feed_author;
+  struct media_file_info mfi = { 0 };
+  struct rss_item_info ri;
+  uint32_t time_added;
+  void *ptr = NULL;
+  int ret;
+
+  xml = rss_xml_get(pli->path);
+  if (!xml)
+    {
+      DPRINTF(E_LOG, L_LIB, "Could not get RSS/xml from '%s' (id %d)\n", pli->path, pli->id);
+      return -1;
+    }
+
+  ret = rss_xml_parse_feed(&feed_title, &feed_author, xml);
+  if (ret < 0)
+    {
+      DPRINTF(E_LOG, L_LIB, "Invalid RSS/xml received from '%s' (id %d)\n", pli->path, pli->id);
+      mxmlDelete(xml);
+      return -1;
+    }
+
+  free(pli->title);
+  pli->title = safe_strdup(feed_title);
+
+  // Fake the time - useful when we are adding a new stream - since the
+  // newest podcasts are added first (the stream is most recent first)
+  // having time_added date which is older on the most recent episodes
+  // makes no sense so make all the dates the same for a singleu update
+  time_added = (uint32_t)time(NULL);
+
+  // Walk through the xml, saving each item
+  *count = 0;
+  db_transaction_begin();
+  while ((ret = rss_xml_parse_item(&ri, xml, &ptr)) == 0 && (*count < pli->query_limit))
+    {
+      if (library_is_exiting())
+	{
+	  db_transaction_rollback();
+	  mxmlDelete(xml);
+	  return -1;
+	}
+
+      db_pl_add_item_bypath(pli->id, ri.url);
+      (*count)++;
+
+      // Try to just ping if already in library
+      if (scan_type == RSS_SCAN_RESCAN)
+	{
+	  ret = db_file_ping_bypath(ri.url, 0);
+	  if (ret > 0)
+	    continue;
+	}
+
+      scan_metadata_stream(&mfi, ri.url);
+
```

Comment:
```
  if (mfi->song_length == 0 && mfi->file_size == 0)
    {
      DPRINTF(E_INFO, L_LIB, "Ignoring item (empty media) RSS id: %d name: '%s' url: %s pubdate: %s title: '%s'\n", pl_id, rss_feed_title, rss_item_url, rss_item_pubDate, rss_item_title);

      free_mfi(&mfi,1);
      continue;
    }
```
I would like to add this check back  - i've seen some elements that for some reason are published to the RSS stream but are actual empty media files.

Reply:
Actually, this is one of the reasons why I went with scanning the underlying media instead of using the info from the RSS for title etc.

Reply:
Can you share the RSS that has these empty files? I'm not sure it is forked-daapd's job to sanitize RSS's, but I suppose it depends on the reason they are empty. Also, having the check could lead to somewhat inconsistent behaviour if the file is unavailable during first time scan. Then it won't be added, but it will be added later if it comes online. Then it will stay there, even if it disappears again.

Reply:
Trying to recall which RSS I saw this empty file but nothing there - may have been a temp issue with ffmpeg not reading from the remote server but I can't reproduce this right now.

However going through my RSS on this dev host, I see `http://cgpgrey.libsyn.com/rss` does have an RSS entry that has no `url` so we'll need
```
index 3a3e4827..d40e0c84 100644
--- a/src/library/rssscanner.c
+++ b/src/library/rssscanner.c
@@ -437,6 +437,12 @@ rss_save(struct playlist_info *pli, int *count, enum rss_scan_type scan_type)
          return -1;
        }
 
+      if (!ri.url)
+        {
+          DPRINTF(E_INFO, L_LIB, "Ignoring item (empty media) RSS id: %d name: '%s' url: %s pubdate: %s title: '%s'\n", pli->id, pli->title, ri.url, ri.pubdate, ri.title);
+          continue;
+        }
+
       db_pl_add_item_bypath(pli->id, ri.url);
       (*count)++;
```
The entry for this empty comes from `Thu, 24 Jan 2013` on this so you'll have to set `RSS_DEFAULT_LIMIT 100` to see this.